src/main/java/org/elasticsearch/indices/store/IndicesStore.java

```diff
@@ -333,8 +336,69 @@ public String executor() {
+            boolean shardActive = shardActive(request);
+            if (shardActive) {
+                channel.sendResponse(new ShardActiveResponse(shardActive(request), clusterService.localNode()));
+            } else {
+                // shard is not active, might be POST_RECOVERY so check if cluster state changed inbetween or wait for next change
+                observer.waitForNextChange(new ClusterStateObserver.Listener() {
+                    @Override
+                    public void onNewClusterState(ClusterState state) {
+                        try {
+                            channel.sendResponse(new ShardActiveResponse(shardActive(request), clusterService.localNode()));
+                        } catch (IOException e) {
+                            logger.error("failed send response for shard active while trying to delete shard {} - shard will probably not be removed", request.shardId, new Exception());
+                            logger.error("Exception: ", e);
+                        }
+                    }
+
+                    @Override
+                    public void onClusterServiceClose() {
+                        try {
+                            channel.sendResponse(new ShardActiveResponse(false, clusterService.localNode()));
+                        } catch (IOException e) {
+                            logger.error("failed send response for shard active while trying to delete shard {} - shard will probably not be removed", request.shardId, new Exception());
+                            logger.error("Exception: ", e);
+                        }
+                    }
+
+                    @Override
+                    public void onTimeout(TimeValue timeout) {
+                        // Try one more time
+                        try {
+                            channel.sendResponse(new ShardActiveResponse(shardActive(request), clusterService.localNode()));
+                        } catch (IOException e) {
+                            logger.error("failed send response for shard active while trying to delete shard {} - shard will probably not be removed", request.shardId, new Exception());
+                            logger.error("Exception: ", e);
+                        }
+                    }
+                }, new ClusterStateObserver.ChangePredicate() {
+                    @Override
+                    public boolean apply(ClusterState previousState, ClusterState.ClusterStateStatus previousStatus,
+                                         ClusterState newState, ClusterState.ClusterStateStatus newStatus) {
+                        return shardActive(request);
+                    }
+
+                    @Override
+                    public boolean apply(ClusterChangedEvent event) {
+                        return shardActive(request);
+                    }
+                });
+            }
         }
 
         private boolean shardActive(ShardActiveRequest request) {
```

Comment:
Should not be a problem. `IndicesClusterStateService` is in the `ClusterService.priorityClusterStateListeners` and the observer is in the `ClusterService.postAppliedListeners` which are notified afterwards so we can be sure the shard has moved to started. In addition, the ClusterState does not tell if the shard is actually moved to started.